tls/s2n_protocol_preferences.c

```diff
@@ -17,39 +17,88 @@
+     */
+    ENSURE(protocol_len != 0, S2N_ERR_INVALID_APPLICATION_PROTOCOL);
+
+    uint32_t prev_len = application_protocols->size;
+    uint32_t new_len = prev_len + /* len prefix */ 1 + protocol_len;
+    ENSURE(new_len <= UINT16_MAX, S2N_ERR_INVALID_APPLICATION_PROTOCOL);
+
+    GUARD_AS_RESULT(s2n_realloc(application_protocols, new_len));
+
     struct s2n_stuffer protocol_stuffer = {0};
+    GUARD_AS_RESULT(s2n_stuffer_init(&protocol_stuffer, application_protocols));
+    GUARD_AS_RESULT(s2n_stuffer_skip_write(&protocol_stuffer, prev_len));
+    GUARD_AS_RESULT(s2n_stuffer_write_uint8(&protocol_stuffer, protocol_len));
+    GUARD_AS_RESULT(s2n_stuffer_write_bytes(&protocol_stuffer, protocol, protocol_len));
 
-    GUARD(s2n_free(application_protocols));
+    return S2N_RESULT_OK;
+}
 
+S2N_RESULT s2n_protocol_preferences_set(struct s2n_blob *application_protocols, const char *const *protocols, int protocol_count)
+{
+    ENSURE_MUT(application_protocols);
+
+    /* NULL value indicates no preference so free the previous blob */
     if (protocols == NULL || protocol_count == 0) {
-        /* NULL value indicates no preference, so nothing to do */
-        return 0;
+        GUARD_AS_RESULT(s2n_free(application_protocols));
+        return S2N_RESULT_OK;
     }
 
-    GUARD(s2n_stuffer_growable_alloc(&protocol_stuffer, 256));
-    for (int i = 0; i < protocol_count; i++) {
+    DEFER_CLEANUP(struct s2n_blob new_protocols = { 0 }, s2n_free);
+
+    for (size_t i = 0; i < protocol_count; i++) {
+        const uint8_t * protocol = (const uint8_t *)protocols[i];
         size_t length = strlen(protocols[i]);
-        uint8_t protocol[255];
 
-        S2N_ERROR_IF(length > 255 || (s2n_stuffer_data_available(&protocol_stuffer) + length + 1) > 65535, S2N_ERR_APPLICATION_PROTOCOL_TOO_LONG);
-        memcpy_check(protocol, protocols[i], length);
-        GUARD(s2n_stuffer_write_uint8(&protocol_stuffer, length));
-        GUARD(s2n_stuffer_write_bytes(&protocol_stuffer, protocol, length));
+        /**
+         *= https://tools.ietf.org/rfc/rfc7301#section-3.1
+         *# Empty strings
+         *# MUST NOT be included and byte strings MUST NOT be truncated.
+         */
+        ENSURE(length < 256, S2N_ERR_INVALID_APPLICATION_PROTOCOL);
+
+        GUARD_RESULT(s2n_protocol_preferences_append(&new_protocols, protocol, (uint8_t)length));
     }
 
-    GUARD(s2n_stuffer_extract_blob(&protocol_stuffer, application_protocols));
-    GUARD(s2n_stuffer_free(&protocol_stuffer));
-    return 0;
+    /* now we can free the previous list since we've validated all new input */
+    GUARD_AS_RESULT(s2n_free(application_protocols));
+
+    /* update the connection/config application_protocols with the newly allocated blob */
+    *application_protocols = new_protocols;
+
+    /* zero out new_protocols so we no longer refer to what we just allocated */
+    new_protocols = (struct s2n_blob){ 0 };
+    GUARD_AS_RESULT(s2n_free(&new_protocols));
```

Comment:
What is the point of combining this with DEFER_CLEANUP? Couldn't you drop both DEFER_CLEANUP and these two lines?

Reply:
Ideally we wouldn't need the `s2n_free` call but cppcheck complains that i'm assigning `new_protocols` and never reading the value (which isn't true in the end, since DEFER_CLEANUP will read it).

Reply:
You might want to add a note about just calling free to make cppcheck happy to the code. Alternatively, it might be more intuitive to use a cppcheck in-line suppression than to give in to its false positives.

Reply:
I've added a note